File: python-scripts/fetch_users.py
import psycopg2
from psycopg2 import sql

def connect_to_db():
    try:
        connection = psycopg2.connect(
            host="localhost",
            database="yourdb",
            user="youruser",
            password="yourpassword"
        )
        return connection
    except psycopg2.Error as e:
        print(f"Error connecting to the database: {e}")
        exit(1)

def fetch_users(connection):
    try:
        with connection.cursor() as cursor:
            query = sql.SQL("SELECT id, username, email, created_at FROM users")
            cursor.execute(query)
            rows = cursor.fetchall()
            if rows:
                # Print column headers
                print(f"{'ID':<5}{'Username':<20}{'Email':<30}{'Created At':<20}")
                print("=" * 75)
                # Print rows
                for row in rows:
                    print(f"{row[0]:<5}{row[1]:<20}{row[2]:<30}{row[3]:<20}")
            else:
                print("No users found.")
    except psycopg2.Error as e:
        print(f"Error executing query: {e}")

def main():
    connection = connect_to_db()
    print("Connected to the database successfully!")
    print("Fetching users:")
    fetch_users(connection)
    connection.close()
    print("Connection closed.")

if __name__ == "__main__":
    main()

File: cpp-components/fetch_users.cpp
#include <iostream>
#include <libpq-fe.h>

void checkConnection(PGconn* conn) {
    if (PQstatus(conn) != CONNECTION_OK) {
        std::cerr << "Connection to database failed: " << PQerrorMessage(conn) << std::endl;
        PQfinish(conn);
        exit(1);
    }
}

void fetchUsers(PGconn* conn) {
    const char* query = "SELECT id, username, email, created_at FROM users";
    PGresult* res = PQexec(conn, query);

    if (PQresultStatus(res) != PGRES_TUPLES_OK) {
        std::cerr << "Failed to execute query: " << PQerrorMessage(conn) << std::endl;
        PQclear(res);
        PQfinish(conn);
        exit(1);
    }

    int rows = PQntuples(res);
    int cols = PQnfields(res);

    // Print column headers
    for (int col = 0; col < cols; ++col) {
        std::cout << PQfname(res, col) << "\t";
    }
    std::cout << std::endl;

    // Print rows
    for (int row = 0; row < rows; ++row) {
        for (int col = 0; col < cols; ++col) {
            std::cout << PQgetvalue(res, row, col) << "\t";
        }
        std::cout << std::endl;
    }

    PQclear(res);
}

int main() {
    const char* conninfo = "host=localhost dbname=yourdb user=youruser password=yourpassword";
    PGconn* conn = PQconnectdb(conninfo);

    checkConnection(conn);

    std::cout << "Connected to the database successfully!" << std::endl;
    std::cout << "Fetching users:" << std::endl;

    fetchUsers(conn);

    PQfinish(conn);
    std::cout << "Connection closed." << std::endl;

    return 0;
}
